File: 2_Singleton.cs
/*
题目：设计一个类，我们只能生成该类的一个实例
扩展：大话设计模式，
注意：自己在纸上书写程序@Qzl

*/
//单线程
class Singleton {
	private Singleton instance;
	private Singleton(){ } //静态构造函数
	public Singleton getInstance(){
		if(instance == null)
			instance = new Singleton();
		return instance;
	}
}
//多线程
class Singleton1 {
	private Singleton1 instance;
	private static readonly object sync = new object();
	private Singleton(){ }
	public Singleton1 getInstance(){
		if(instance == null){ //提高效率，没必要每次试图得到实例时都加入同步锁
			lock(sync){ //加同步锁
				if(instance == null)
					instance = new Singleton1();
			}
		}
		return instance;
	}
}
//C#静态初始化方法
class sealed class Singleton2{
	private static readonly Singleton2 instance = new Singleton2();
	private Singleton2(){}
	public Singleton2 getInstance(){
		return instance;
	}
}

File: 2_Singleton.cpp
#include<iostream>
using namespace std;
/*
题目：设计一个类，我们只能生成该类的一个实例
扩展：大话设计模式，
注意：自己在纸上书写程序@Qzl

*/
//C++版本
//单线程
class Singleton {
public:
	Singleton* getInstance(){
		if(instance == NULL)
			instance = new Singleton();
		return instance;
	}
private:
	Singleton(){ } //静态构造函数
	static Singleton *instance;
};
//多线程
class Singleton1 {
public:
	Singleton1* getInstance(){
		lock();
		if(instance == NULL)
			instance = new Singleton1();
		unlock();
		return instance;
	}
private:
	Singleton(){ } //静态构造函数
	static Singleton1 *instance;
};

class Singleton2{
public:
	Singleton2* getInstance(){
		if(instance == NULL){
			lock();
			if(instance == NULL)
				instance = new Singleton2();
			unlock();
		}
		return instance;
	}
private:
	Singleton(){ } //静态构造函数
	static Singleton2 *instance;
};
